File: src/main/java/com/qtech/bubbles/media/AudioSlot.java
package com.qtech.bubbles.media;

import javafx.scene.media.AudioClip;

import java.io.File;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;

public class AudioSlot {
    private final AudioClip clip;
    private final String name;
    private boolean stopped = false;

    public AudioSlot(File file) {
        this(file, "");
    }

    public AudioSlot(File file, String name) {
        clip = new AudioClip(file.toURI().toString());
        this.name = name;
    }

    public AudioSlot(URI uri) {
        this(uri, "");
    }

    public AudioSlot(URI uri, String name) {
        clip = new AudioClip(uri.toString());
        this.name = name;
    }

    public AudioSlot(URL url) throws URISyntaxException {
        this(url, "");
    }

    public AudioSlot(URL url, String name) throws URISyntaxException {
        clip = new AudioClip(url.toURI().toString());
        this.name = name;
    }

    public AudioClip getClip() {
        return clip;
    }

    public void play() {
        clip.play();
        stopped = false;
    }

    public void stop() {
        clip.stop();
        stopped = true;
    }

    public double getBalance() {
        return clip.getBalance();
    }

    public double getRate() {
        return clip.getRate();
    }

    public void setVolume(double v) {
        clip.setVolume(v);
    }

    public void setMute(double v) {
        clip.setBalance(v);
    }

    public double getVolume() {
        return clip.getVolume();
    }

    public String getSource() {
        return clip.getSource();
    }

    public boolean isStopped() {
        return stopped;
    }

    public boolean isPlaying() {
        return !stopped;
    }

    public String getName() {
        return name;
    }
}

File: src/main/kotlin/qtech/bubbles/media/AudioSlot.kt
package qtech.bubbles.media

import javafx.scene.media.AudioClip
import java.io.File
import java.net.URI
import java.net.URL

class AudioSlot {
    val clip: AudioClip
    val name: String
    var isStopped = false
        private set

    constructor(file: File, name: String = "") {
        clip = AudioClip(file.toURI().toString())
        this.name = name
    }

    constructor(uri: URI, name: String = "") {
        clip = AudioClip(uri.toString())
        this.name = name
    }

    constructor(url: URL, name: String = "") {
        clip = AudioClip(url.toURI().toString())
        this.name = name
    }

    fun play() {
        clip.play()
        isStopped = false
    }

    fun stop() {
        clip.stop()
        isStopped = true
    }

    val balance: Double
        get() = clip.balance
    val rate: Double
        get() = clip.rate

    fun setMute(v: Double) {
        clip.balance = v
    }

    var volume: Double
        get() = clip.volume
        set(v) {
            clip.volume = v
        }
    val source: String
        get() = clip.source
    val isPlaying: Boolean
        get() = !isStopped
}